core/include/scipp/core/multi_index.h

```diff
@@ -310,24 +383,188 @@ template <scipp::index N> class MultiIndex {
+                          const scipp::index end_index) {
+    scipp::index res = 0;
+    for (; begin_index < end_index; ++begin_index) {
+      res += coord(begin_index) * stride(begin_index, i_data);
+    }
+    return res;
+  }
+
+  /// Stride for each operand in each dimension.
+  [[nodiscard]] scipp::index &stride(const scipp::index dim,
+                                     const scipp::index data) noexcept {
+    return m_buffer[data + dim * N];
+  }
+
+  [[nodiscard]] const scipp::index &
+  stride(const scipp::index dim, const scipp::index data) const noexcept {
+    return m_buffer[data + dim * N];
+  }
+
+  template <size_t... I, class... StridesArgs>
+  void store_strides_impl(const scipp::index ndim, const scipp::index first_dim,
+                          std::index_sequence<I...>,
+                          const StridesArgs &... strides) noexcept {
+    (
+        [&]() {
+          for (scipp::index dim = 0; dim < ndim; ++dim) {
+            stride(first_dim + dim, I) = strides[ndim - 1 - dim];
+          }
+        }(),
+        ...);
+  }
+
+  template <class... StridesArgs>
+  void store_strides(const scipp::index ndim, const scipp::index first_dim,
+                     const StridesArgs &... strides) noexcept {
+    store_strides_impl(ndim, first_dim,
+                       std::index_sequence_for<StridesArgs...>(), strides...);
+  }
+
+  /// Current index in iteration dimensions for both bin and inner dims.
+  [[nodiscard]] scipp::index &coord(const scipp::index dim) noexcept {
+    return *coord_it(dim);
+  }
+
+  [[nodiscard]] const scipp::index &
+  coord(const scipp::index dim) const noexcept {
+    return *coord_it(dim);
+  }
+
+  [[nodiscard]] auto coord_it(const scipp::index dim = 0) const noexcept {
+    return m_buffer.get() + (std::max(m_ndim, scipp::index{2}) * N + dim);
+  }
+
+  [[nodiscard]] auto coord_end() const noexcept {
+    return m_buffer.get() + (std::max(m_ndim, scipp::index{2}) * (N + 1));
+  }
+
+  /// Shape of the iteration dimensions for both bin and inner dims.
+  [[nodiscard]] scipp::index &shape(const scipp::index dim) noexcept {
+    return *shape_it(dim);
+  }
+
+  [[nodiscard]] const scipp::index &
+  shape(const scipp::index dim) const noexcept {
+    return *shape_it(dim);
+  }
+
+  [[nodiscard]] auto shape_it(const scipp::index dim = 0) const noexcept {
+    return m_buffer.get() + (std::max(m_ndim, scipp::index{2}) * (N + 1) + dim);
+  }
+
+  [[nodiscard]] auto shape_end() const noexcept {
+    return m_buffer.get() + (std::max(m_ndim, scipp::index{2}) * (N + 2));
+  }
+
+  [[nodiscard]] auto copy_buffer() const {
+    const auto size = detail::get_buffer_size<N>(m_ndim);
+    auto new_buffer = std::make_unique<scipp::index[]>(size);
+    std::copy(m_buffer.get(), m_buffer.get() + size, new_buffer.get());
+    return new_buffer;
+  }
+
+  /// Container for strides, coords, shape.
+  std::unique_ptr<scipp::index[]> m_buffer;
```

Comment:
Unless you have a good reason why this *has* to be dynamic, I would still just use a buffer on the stack (i.e., as part of this class). It is very hard to predict what the performance of memory allocations is in different contexts and on different systems, so unless you see a clear performance *advantage*, avoid it.

Reply:
It is not about performance but about lifting the limit of having at most 6 dimensions.

Reply:
Isn't it still limited by `class Dimensions`?

Reply:
True, we would need to do this change in several places.